### PR TITLE
Change to propagate the last Response while retry in some condition

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
When the number of attempts exceeds max total attempts or the calculated next delay is
after the moment which response timeout will be happen, `RetryingClient` is finished
with `RetryGiveUpException` or `ResponseTimeoutException`.
However these Exceptions are meaningless and if a user can get the last response from the
server instead, it will be much helpful to figure out what's going on.

Modifications:
- Change to propagate the last `Response` in above condition
- Delele `RetryGiveUpException` class completely
- Refactor test codes

Result:
- Robustness

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java [hunk 1]
```diff
@@ -122,10 +130,8 @@ protected final boolean setResponseTimeout(ClientRequestContext ctx) {
      *
      * <p>{@code Math.min(responseTimeoutMillis, Backoff.nextDelayMillis(int))}
      *
-     * @return the number of milliseconds to wait for before attempting a retry
-     * @throws RetryGiveUpException if the current attempt number is greater than {@code maxTotalAttempts} or
-     *                              {@link Backoff#nextDelayMillis(int)} returns -1
-     * @throws ResponseTimeoutException if the remaining response timeout is equal to or less than 0
+     * @return the number of milliseconds to wait for before attempting a retry. -1 if the currentAttemptNo
```
Comment (hyangtack) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java]: `{@code currentAttemptNo}`, `{@code maxAttempts}`, `{@code nextDelay}`?

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java [hunk 2]
```diff
@@ -90,6 +94,10 @@ protected final O executeDelegate(ClientRequestContext ctx, I req) throws Except
         }
     }
 
+    protected static Logger logger() {
+        return logger;
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java]: I'm not sure we really need this. A subclass should use its own `Logger`? e.g. `RetryingHttpClient` should use the `Logger` with the name `RetryingHttpClient`, not `RetryingClient`.

File: site/src/sphinx/client-retry.rst [hunk 3]
```diff
@@ -211,7 +214,27 @@ it is timed out by the per-attempt timeout.
                                            10,000ms (ResponseTimeoutException)
     @endditaa
 
+In above example, every attempt is made before it is timed out because the :api:`Backoff` is disabled.
```
Comment (trustin) [inline site/src/sphinx/client-retry.rst]: - In the example above

File: site/src/sphinx/client-retry.rst [hunk 4]
```diff
@@ -211,7 +214,27 @@ it is timed out by the per-attempt timeout.
...
+However, What if a :api:`Backoff` is enabled and the moment of trying next attempt is after the point of
```
Comment (trustin) [inline site/src/sphinx/client-retry.rst]: What if -> what if

File: site/src/sphinx/client-retry.rst [hunk 5]
```diff
@@ -211,7 +214,27 @@ it is timed out by the per-attempt timeout.
...
+occurring :api:`ResponseTimeoutException`? The :api:`RetryingClient` do not schedule for the next attempt and
```
Comment (trustin) [inline site/src/sphinx/client-retry.rst]: - 'occurring' could be dropped?
- The RetryingClient -> In such a case, the RetryingClient
- do not -> does not

File: site/src/sphinx/client-retry.rst [hunk 6]
```diff
@@ -211,7 +214,27 @@ it is timed out by the per-attempt timeout.
...
+returned with the latest :api:`Response` from the server immediately. Consider following example:
```
Comment (trustin) [inline site/src/sphinx/client-retry.rst]: - returned with the latest ..immediately. -> finish the retry session immediately with the last ...
- following example -> the following example

File: site/src/sphinx/client-retry.rst [hunk 7]
```diff
@@ -211,7 +214,27 @@ it is timed out by the per-attempt timeout.
...
+    0ms         3,000ms     6,000ms     9,000ms     12,000ms
+    |           |           |           |           |
+    +-----------+-----------+-----------+-----------+--------------------+
+    | Attempt 1 |           | Attempt 2 |           | Attempt 3 is made? |
+    +-----------+-----------+-----------+----+------+--------------------+
+                                             |
+                                           10,000ms (ResponseTimeoutException)
+    @endditaa
 
+Unlike above example, the :api:`Backoff` is enabled and it makes the :api:`RetryingClient` retries after ``3``
```
Comment (trustin) [inline site/src/sphinx/client-retry.rst]: - the example above
- retries after 3 seconds -> perform retries with 3-second delays

File: site/src/sphinx/client-retry.rst [hunk 8]
```diff
@@ -211,7 +214,27 @@ it is timed out by the per-attempt timeout.
...
+seconds. When the second attempt is finished, the next attempt will be 12,000ms which is after 10,000ms.
```
Comment (trustin) [inline site/src/sphinx/client-retry.rst]: - finished -> finished at 9,000ms
- 12,000ms -> at 12,000ms
- which is after 10,000ms -> exceeding the response timeout of 10,000ms

File: site/src/sphinx/client-retry.rst [hunk 9]
```diff
@@ -211,7 +214,27 @@ it is timed out by the per-attempt timeout.
...
+The :api:`RetryingClient`, at this point, stops to retry and returned with the latest :api:`Response` which it
```
Comment (trustin) [inline site/src/sphinx/client-retry.rst]: - stops to retry -> stops retrying
- returned with the latest Response -> finishes the retry session with the last Response

File: site/src/sphinx/client-retry.rst [hunk 10]
```diff
@@ -211,7 +214,27 @@ it is timed out by the per-attempt timeout.
...
+gets from the server at 9,000ms.
```
Comment (trustin) [inline site/src/sphinx/client-retry.rst]: which it gets from the server at -> , retrieved at

File: site/src/sphinx/client-retry.rst [hunk 11]
```diff
@@ -170,6 +170,9 @@ the maximum number of total attempts to 10 by default. You can change this value
 Or, you can override the default value of 10 using the JVM system property
 ``-Dcom.linecorp.armeria.defaultMaxTotalAttempts=<integer>``.
 
+Note that when a :api:`RetryingClient` stops by the attempts limit, the client will get the latest
```
Comment (trustin) [inline site/src/sphinx/client-retry.rst]: - by -> due to
- will get the last received Response ?

File: site/src/sphinx/client-retry.rst [hunk 12]
```diff
@@ -211,7 +214,31 @@ it is timed out by the per-attempt timeout.
                                            10,000ms (ResponseTimeoutException)
     @endditaa
 
+In the example above, every attempt is made before it is timed out because the :api:`Backoff` is disabled.
+However, what if a :api:`Backoff` is enabled and the moment of trying next attempt is after the point of
+:api:`ResponseTimeoutException`? In such a case, the :api:`RetryingClient` does not schedule for the
+next attempt and finish the retry session immediately with the last :api:`Response`.
```
Comment (trustin) [inline site/src/sphinx/client-retry.rst]: the last Response -> the last received Response

File: site/src/sphinx/client-retry.rst [hunk 13]
```diff
@@ -211,7 +214,31 @@ it is timed out by the per-attempt timeout.
...
+                                        | 10,000ms (retry session deadline)
+                                        |
+                                    stops retrying at this point
+    @endditaa
 
+Unlike the example above, the :api:`Backoff` is enabled and it makes the :api:`RetryingClient` perform retries
+with 3-second delay. When the second attempt is finished at 9,000ms, the next attempt will be at 12,000ms
+exceeding thre response timeout of 10,000ms.
+The :api:`RetryingClient`, at this point, stops retrying and finished the retry session with the last
+:api:`Response`, retrieved at 9,000ms.
```
Comment (trustin) [inline site/src/sphinx/client-retry.rst]: retrived at 9,000ms from the attempt 2. ?

**Comment by minwoox** [comment]
Thank you~! All fixed!

File: site/src/sphinx/client-retry.rst [hunk 14]
```diff
@@ -211,7 +214,31 @@ it is timed out by the per-attempt timeout.
...
+    +-----------+-----------+-----------+----+------+-----------------------+
+                                        |    |
+                                        | 10,000ms (retry session deadline)
+                                        |
+                                    stops retrying at this point
+    @endditaa
 
+Unlike the example above, the :api:`Backoff` is enabled and it makes the :api:`RetryingClient` perform retries
+with 3-second delay. When the second attempt is finished at 9,000ms, the next attempt will be at 12,000ms
+exceeding thre response timeout of 10,000ms.
```
Comment (hyangtack) [inline site/src/sphinx/client-retry.rst]: thre -> the?

File: site/src/sphinx/client-retry.rst (hunk 12 above)
Comment (hyangtack) [inline site/src/sphinx/client-retry.rst]: finish -> finishes?

File: site/src/sphinx/client-retry.rst (hunk 14 above)
Comment (minwoox) [reply site/src/sphinx/client-retry.rst]: Fixed. Thank you :)

**Comment by minwoox** [comment]
Thank for reviewing~!

**Pull request merged**